docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,216 @@
+
+- backward compatible: Any changes should not introduce backward compatibility issues across at least last two minor leases from the end-user perspective. Using the improved storage system with existing etcd3 backend should just work.
+
+## Direction
+
+We could evolve the current storage layer towards the goal we listed above. It is doable, but it might involve significant interface and implement changes. Also it requires continuous changes of the consumers of the current storage API. During this process, we also need to ensure everything works perfectly. There is no alpha, or beta since we will be changing the single critical path. This is risky, time consuming, and hard to revert.
+
+The solution we proposed here is to start a storage layer from scratch. As Daniel Smith [mentioned](https://github.com/kubernetes/kubernetes/pull/28508#issuecomment-232508836)
+
+> Just to (probably unnecessarily) reiterate my position, I think we're going to need side-by-side implementations of both old and new storage styles, so we shouldn't end up locked into whatever we do now.
+
+We also believe we should start a side-by-side storage implementation to reach the goal we listed. That way, we won’t be affected by any existing issues, and it gives us the opportunities to rethink the design from the ground up, to achieve the goals more clearly. The only risk is that the effort will be completely wasted, if we cannot reach the goals.
+
+## Proposed Storage API
+
+```go
+
+// Storage stores and persists objects.
+// Storage provides read-write access to the objects based on string key.
+type Storage interface {
+	// Wait waits until the storage has seen any version >= given version.
+	// If version == 0, it provides a guarantee that once returning reads should see all writes from this interface.
+	Wait(ctx context.Context, version int64)
+
+	// Put puts an object with a key, and returns current object on given key.
+	// If conditions is not nil, it will do atomic Put. If compare failed, it will return existing object on given key.
+	// See "Conditions" docs for more details on compare logic and returned error.
+	Put(ctx context.Context, key StorageKey, obj runtime.Object, conditions *Conditions) (*KeyResponse, error)
+
+	// Delete deletes a key and its object, and returns deleted object on given key.
+	// conditions has the same guarantee as Put().
+	Delete(ctx context.Context, key StorageKey, conditions *Conditions) (*KeyResponse, error)
+
+	// Get gets the current object of given key.
+	// If version = 0, it will serve data from local cache.
+	// If version > 0, it will get the current state of the key at the time given version is committed.
+	// If no object exists on the key, it will return not found error.
+	Get(ctx context.Context, key StorageKey, version int64) (*KeyResponse, error)
```

Comment:
Is there a way for me to get the current value from the storage system?  I may not care about resourceVersion, but I do want you to attempt to get the latest.

Similar question, is there a way for me to force a quorum read?